iroh/src/endpoint.rs

```diff
@@ -1257,16 +1257,24 @@ impl Connecting {
     }
 
     /// Extracts the ALPN protocol from the peer's handshake data.
-    // Note, we could totally provide this method to be on a Connection as well.  But we'd
-    // need to wrap Connection too.
-    pub async fn alpn(&mut self) -> Result<Vec<u8>> {
+    pub async fn alpn(&mut self) -> Result<Vec<u8>, ConnectionError> {
         let data = self.handshake_data().await?;
-        match data.downcast::<quinn::crypto::rustls::HandshakeData>() {
-            Ok(data) => match data.protocol {
-                Some(protocol) => Ok(protocol),
-                None => bail!("no ALPN protocol available"),
-            },
-            Err(_) => bail!("unknown handshake type"),
+        let data = data
+            .downcast::<quinn::crypto::rustls::HandshakeData>()
+            .expect("fixed crypto setup for iroh");
```

Comment:
(similar to the other comment - at this point we haven't verified that the handshake data is there for sure, so we should probably return an error if it isn't. IMO a CONNECTION_REFUSED is totally fine for this case as well.)